From dcac199af16c23b2391d614eab210233428faf15 Mon Sep 17 00:00:00 2001
From: ian <ian.laffey@mail.mcgill.ca>
Date: Wed, 13 May 2026 22:09:17 +0300
Subject: [PATCH 1/4] AP_NavEKF3: allow init without GPS lock when
 ExtPosCanBootstrap is set

Plane EKF3's InitialiseFilterBootstrap unconditionally refused to start
without a 3D GPS fix, leaving statesInitialised=false and every downstream
flag cleared. With EK3_OPTIONS bit ExtPosCanBootstrap set, allow the
filter to initialise from IMU (tilt) and compass (yaw) alone so attitude
and LOCAL_POSITION_NED at origin are available before the in-air
external-position bootstrap.
---
 libraries/AP_NavEKF3/AP_NavEKF3_core.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/libraries/AP_NavEKF3/AP_NavEKF3_core.cpp b/libraries/AP_NavEKF3/AP_NavEKF3_core.cpp
index 23fd007a1b..72496cafd8 100644
--- a/libraries/AP_NavEKF3/AP_NavEKF3_core.cpp
+++ b/libraries/AP_NavEKF3/AP_NavEKF3_core.cpp
@@ -474,8 +474,12 @@ bool NavEKF3_core::InitialiseFilterBootstrap(void)
     // update sensor selection (for affinity)
     update_sensor_selection();
 
-    // If we are a plane and don't have GPS lock then don't initialise
-    if (assume_zero_sideslip() && dal.gps().status(preferred_gps) < AP_DAL_GPS::GPS_OK_FIX_3D) {
+    // If we are a plane and don't have GPS lock then don't initialise -
+    // unless ExtPosCanBootstrap is set, in which case we initialise from IMU
+    // (tilt) and compass (yaw) alone so the EKF can produce attitude and
+    // LOCAL_POSITION_NED at origin before the in-air bootstrap.
+    if (assume_zero_sideslip() && dal.gps().status(preferred_gps) < AP_DAL_GPS::GPS_OK_FIX_3D &&
+        !(frontend->_options & (int32_t)NavEKF3::Options::ExtPosCanBootstrap)) {
         dal.snprintf(prearm_fail_string,
                      sizeof(prearm_fail_string),
                      "EKF3 init failure: No GPS lock");
-- 
2.39.2 (Apple Git-143)

